cryptocurrency/crypto.py

```diff
@@ -0,0 +1,77 @@
+from pycoingecko import CoinGeckoAPI
+
+from deephaven.DBTimeUtils import secondsToTime, millisToTime
+from deephaven.TableTools import merge
+from deephaven import DynamicTableWriter
+from deephaven import TableTools
+from typing import Callable
+
+import deephaven.Types as dht
+import jpy
+
+from time import sleep, time
+import pandas as pd
+import threading
+
+
+ApplicationState = jpy.get_type("io.deephaven.appmode.ApplicationState")
+
+def crypto_start(app: ApplicationState):
+
+    # minutes to query crypto prices
+    timeToWatch = 2
+
+    # secondsToSleep should be 10 or higher. If too fast, will hit request limit.
+    secondsToSleep = 10
+
+
+    getHistory = False
+
+    # if getHistory = true, the days to pull
+    daysHistory = 90
+
+    # coins to get data
+    ids=['bitcoin', 'ethereum','litecoin', 'dogecoin', 'tether', 'binancecoin', 'cardano', 'ripple', 'polkadot']
+
+    # array to store tables for current and previous data
+    tableArray=[]
+
+    tableWriter = DynamicTableWriter(["Coin", "Timestamp", "Price", "MarketCap", "TotalVolume"], [dht.string, dht.datetime, dht.double, dht.double, dht.double])
+
+
+    tableArray.append(tableWriter.getTable())
+
+    cg = CoinGeckoAPI()
+
+    # get historical data
+    if getHistory:
+        for names in ids:
+            coin_data_hist = cg.get_coin_market_chart_by_id(names, vs_currency = "usd", days = daysHistory)
+            sub = pd.DataFrame(coin_data_hist)
+            tableArray.append(dataFrameToTable(sub).view("Timestamp = millisToTime((long)prices_[i][0])", "Coin = names", "Price = prices_[i][1]", "MarketCap = market_caps_[i][1]", "TotalVolume = total_volumes_[i][1]").moveUpColumns("Timestamp", "Coin"))
+
+    #add each coin data to the master table
+    result = merge(tableArray).selectDistinct("Timestamp", "Coin", "Price", "MarketCap", "TotalVolume").sortDescending("Timestamp", "Coin")
+
+    def thread_func():
+
+        cg.get_coins_markets(vs_currency = 'usd')
+
+        for i in range(int(timeToWatch*60/secondsToSleep)):
+            coin_data = cg.get_price(ids, vs_currencies = 'usd', include_market_cap = True, include_24hr_vol = True, include_24hr_change = False, include_last_updated_at = True)
+
+            for id in ids:
+                #Add new data to the dynamic table
+                tableWriter.logRow( id, secondsToTime(int(coin_data[id]['last_updated_at'])), float(coin_data[id]['usd']), coin_data[id]['usd_market_cap'], coin_data[id]['usd_24h_vol'])
+
+            sleep(secondsToSleep)
+
+    thread = threading.Thread(target = thread_func)
+    thread.start()
+
+
+def initialize(func: Callable[[ApplicationState], None]):
+  app = jpy.get_type("io.deephaven.appmode.ApplicationContext").get()
+  func(app)
+
+initialize(crypto_start)
```

Comment:
So, there are two modes when creating an application script. You can do a "proper" application mode, like you are setting up here - but then any tables you want to expose you need to explicitly set them:

```python
app.setField("crypto_result", result)
```

The alternative is the "global scope" mode, where everything created outside a function implicitly put into the global scope. In that mode, you don't have to worry about ApplicationState/ApplicationContext.

I might suggest: use the same table creation logic you already have, but have the function take your input variables:

```python
def get_coingecko_table(time_to_watch=2, seconds_to_sleep=10, ...):
   ...
   return result
```

After that, you can either do:
```python
crypto_result = get_coingecko_table(...)   
```
for "global mode" or
```python
app = jpy.get_type("io.deephaven.appmode.ApplicationContext").get()
app.setField("crypto_result",  get_coingecko_table(...))
```
for tighter application mode.

Reply:
@devinrsmith 
In the current state of the code the tables are not usable in the session. 
```python
small = HistoricalCryptoTable.where("Price >1")
```
gives `Value: name 'HistoricalCryptoTable' is not defined`

Also the table name is not display name.  Would switching to the "global mode" allow for using the table in the session?
<img width="346" alt="2021-10-01_07-14-15" src="https://user-images.githubusercontent.com/51727488/135610983-1db22635-ebb8-4746-8293-32b983e330b1.png">